File: experiments-local/experiment-17/runs/agent=omp_language=python_model=openrouter/z-ai/glm-5.2_tooling=none/rep2/brazilian_soccer/normalize.py
"""Normalisation helpers for Brazilian soccer data.

CONTEXT
-------
The bundled CSV files come from five different Kaggle datasets and use
inconsistent conventions for the same entities:

* ``Palmeiras-SP`` (Brasileirão), ``América - MG`` (Copa do Brasil) and
  ``Barcelona-EQU`` (Libertadores) carry a trailing state/country suffix.
* The extended-stats file uses bare names (``Palmeiras``) for unambiguous
  clubs but spelled-out disambiguators (``Atletico Mineiro``,
  ``Botafogo RJ``) for ambiguous ones.
* Accents are kept in some files (``São Paulo``) and dropped in others
  (``Sao Paulo``).

Crucially, several short names are shared by *different* clubs that are
only disambiguated by their state — ``Atlético-MG`` (Atlético Mineiro)
vs ``Atlético-GO`` (Atlético Goianiense) vs ``Athletico-PR`` (Athletico
Paranaense).  Stripping the suffix would wrongly merge them, so the
canonical key **retains** the state token.  Equivalence between the
state-suffixed form (``flamengo rj``) and the bare form (``flamengo``)
is resolved through an alias map built in :func:`build_alias_map`.

Two names refer to the same team iff ``resolve_team_key`` agrees::

    resolve_team_key(map, "Palmeiras-SP") == resolve_team_key(map, "Palmeiras")
    resolve_team_key(map, "Atlético-MG") == resolve_team_key(map, "Atletico Mineiro")
    resolve_team_key(map, "Atlético-MG") != resolve_team_key(map, "Atletico-GO")
"""

from __future__ import annotations

import re
import unicodedata
from datetime import datetime
from typing import Iterable

_BR_STATE_WORDS = {
    "ac", "al", "ap", "am", "ba", "ce", "df", "es", "go", "ma",
    "mt", "ms", "mg", "pa", "pb", "pr", "pe", "pi", "rj", "rn",
    "rs", "ro", "rr", "sc", "sp", "se", "to",
}

# Trailing "-XX" / " - XX" state or country suffix (2-3 uppercase letters).
_SUFFIX_RE = re.compile(r"\s*-\s*([A-Z]{2,3})\s*$")
# Parenthetical notes: "Nacional (URU)", "Time (antigo ...)".
_PAREN_RE = re.compile(r"\s*\([^)]*\)")


def strip_accents(text: str) -> str:
    """Return *text* with combining diacritics removed (ASCII fold)."""
    nfkd = unicodedata.normalize("NFKD", text)
    return "".join(ch for ch in nfkd if not unicodedata.combining(ch))


_SMALL_WORDS = {"da", "de", "do", "dos", "das", "e", "del", "la"}


def titlecase_name(name: str) -> str:
    """Title-case a team name, preserving acronyms (``CSA``) and lowering
    Portuguese connectors (``da``, ``de``)."""
    words = str(name).split()
    out = []
    for i, w in enumerate(words):
        if w.isupper() and len(w) <= 4:
            out.append(w)
        elif w and w[0].isdigit():
            out.append(w)
        elif w.lower() in _SMALL_WORDS and i > 0:
            out.append(w.lower())
        else:
            out.append(w[:1].upper() + w[1:].lower())
    return " ".join(out)


def _split_suffix(s: str) -> tuple[str, str | None]:
    """Return (body, state) where *state* is the trailing suffix code or
    None.  Operates on the original-cased string so the [A-Z] class works."""
    m = _SUFFIX_RE.search(s)
    if m:
        return s[: m.start()].strip(), m.group(1)
    return s.strip(), None


# Common Brazilian club legal-form abbreviations that appear as a leading
# or trailing token ("EC Bahia", "Fortaleza FC", "4 de Julho EC").  They
# are stripped so the underlying club name matches across sources.
_CLUB_TOKENS = {"EC", "FC", "SC", "AC", "AA", "GE", "GR", "SE", "CE", "CA"}


def _strip_club_tokens(body: str) -> str:
    """Drop leading/trailing club legal-form tokens (EC, FC, SC, ...)."""
    parts = body.split()
    while parts and parts[0].upper() in _CLUB_TOKENS:
        parts.pop(0)
    while parts and parts[-1].upper() in _CLUB_TOKENS:
        parts.pop()
    return " ".join(parts)


def has_state_suffix(name) -> bool:
    """True if *name* carries a trailing ``-XX`` state/country suffix."""
    if name is None:
        return False
    return bool(_SUFFIX_RE.search(str(name)))


def canonical_key(name) -> str:
    """Reduce a team/club name to a stable matching key that RETAINS the
    state token.

    Examples::

        canonical_key("Palmeiras-SP")  == "palmeiras sp"
        canonical_key("América - MG")  == "america mg"
        canonical_key("Barcelona-EQU") == "barcelona equ"
        canonical_key("São Paulo")     == "sao paulo"
        canonical_key("Botafogo RJ")   == "botafogo rj"
    """
    if name is None:
        return ""
    s = _PAREN_RE.sub("", str(name)).strip()
    body, state = _split_suffix(s)
    body = strip_accents(body).lower().strip()
    body = _strip_club_tokens(body)
    body = re.sub(r"\s+", " ", body).strip()
    if state:
        return f"{body} {state.lower()}"
    return body


def bare_key(name) -> str:
    """Canonical key with the state suffix stripped (used to detect
    ambiguity between same-named clubs from different states)."""
    if name is None:
        return ""
    s = _PAREN_RE.sub("", str(name)).strip()
    body, _ = _split_suffix(s)
    body = strip_accents(body).lower().strip()
    body = _strip_club_tokens(body)
    return re.sub(r"\s+", " ", body).strip()


def display_name(name) -> str:
    """Return a clean display name: accents kept, suffix/notes dropped,
    and a trailing bare state word (``Vasco da Gama RJ``) removed."""
    if name is None:
        return ""
    s = _PAREN_RE.sub("", str(name)).strip()
    body, _ = _split_suffix(s)
    body = _strip_club_tokens(body)
    body = re.sub(r"\s+", " ", body).strip()
    # Drop a trailing bare 2-letter Brazilian state word, e.g. "RJ".
    parts = body.split()
    if len(parts) >= 2 and parts[-1].lower() in _BR_STATE_WORDS and len(parts[-1]) == 2:
        body = " ".join(parts[:-1])
    return titlecase_name(body)


# --- Alias resolution -----------------------------------------------------

# Curated map of spelled-out / bare ambiguous forms -> state-retaining id.
# These cover the well-known Brazilian clubs whose names collide across
# states and whose extended-stats spellings differ from the suffix form.
_CURATED_ALIASES = {
    # Spelled-out forms (extended-stats dataset) -> state id.
    "atletico mineiro": "atletico mg",
    "atletico goianiense": "atletico go",
    "atletico paranaense": "atletico pr",
    "athletico paranaense": "atletico pr",
    "athletico pr": "atletico pr",
    "atletico acreano": "atletico ac",
    "atletico alagoinhas": "atletico ba",
    "sport recife": "sport pe",
    "america fc natal": "america rn",
    "red bull bragantino": "bragantino sp",
    "vasco": "vasco da gama rj",
    # Bare ambiguous names -> the historically prominent club, so that a
    # user typing just "Botafogo" or "Santos" resolves to the famous one.
    "botafogo": "botafogo rj",
    "santos": "santos sp",
    "internacional": "internacional rs",
    "santa cruz": "santa cruz pe",
    "juventude": "juventude rs",
    "bragantino": "bragantino sp",
    "america": "america mg",
    "atletico": "atletico mg",
    "athletico": "atletico pr",
    "vitoria": "vitoria ba",
    "flamengo": "flamengo rj",
    "guarani": "guarani sp",
    "operario": "operario ms",
    "rio branco": "rio branco ac",
    "sao raimundo": "sao raimundo pa",
    "ypiranga": "ypiranga rs",
    "comercial": "comercial ms",
    "abc": "abc rn",
}

# Display-name overrides for ids whose most common spelling is an ambiguous
# bare name (e.g. "Atlético"); use the full disambiguated name instead.
DISPLAY_OVERRIDES = {
    "atletico mg": "Atlético Mineiro",
    "atletico pr": "Athletico Paranaense",
    "atletico go": "Atlético Goianiense",
    "atletico ac": "Atlético Acreano",
    "atletico ba": "Atlético de Alagoinhas",
    "atletico es": "Atlético Esporte Clube",
    "america mg": "América Mineiro",
    "america rn": "América Natal",
    "sport pe": "Sport Recife",
}


def build_alias_map(names: Iterable) -> dict[str, str]:
    """Build an alias map from the set of raw team *names*.

    The map resolves any spelling to a state-retaining canonical id:

    * every state-suffixed name maps to itself,
    * a bare name that corresponds to exactly ONE state-suffixed club is
      mapped to that club's id (auto-derivation, e.g. ``palmeiras`` ->
      ``palmeiras sp``),
    * curated entries override for spelled-out / ambiguous cases.
    """
    alias_map: dict[str, str] = {}
    by_bare: dict[str, list[str]] = {}

    for name in names:
        if name is None:
            continue
        s = str(name).strip()
        if not s:
            continue
        if has_state_suffix(s):
            sid = canonical_key(s)
            alias_map.setdefault(sid, sid)
            by_bare.setdefault(bare_key(s), []).append(sid)
        else:
            # Non-suffixed spelling: identity unless a curated/auto entry
            # resolves it later.
            k = canonical_key(s)
            alias_map.setdefault(k, k)

    # Auto-derive bare -> state id for unambiguous clubs.
    for bare, sids in by_bare.items():
        unique = sorted(set(sids))
        if len(unique) == 1:
            alias_map[bare] = unique[0]

    # Curated overrides (applied last so they win).
    alias_map.update(_CURATED_ALIASES)
    return alias_map


def resolve_team_key(alias_map: dict[str, str], name) -> str:
    """Resolve a team name (raw or user-typed) to its canonical id."""
    key = canonical_key(name)
    if key in alias_map:
        return alias_map[key]
    # Fall back to the bare form (e.g. user typed "Atletico Mineiro" which
    # canonical_key keeps as-is; curated map covers the common cases).
    bk = bare_key(name)
    return alias_map.get(bk, key)


# --- Date parsing ---------------------------------------------------------

_DATE_FORMATS = ("%Y-%m-%d %H:%M:%S", "%Y-%m-%d", "%d/%m/%Y")


def parse_date(value) -> datetime | None:
    """Parse the date/datetime strings found across the datasets.

    Handles ISO datetimes (``2012-05-19 18:30:00``), plain ISO dates
    (``2023-09-24``) and Brazilian ``DD/MM/YYYY`` dates.  Returns ``None``
    for missing/unparseable input.
    """
    if value is None:
        return None
    s = str(value).strip()
    if not s or s.lower() in {"nan", "none", "nat"}:
        return None
    for fmt in _DATE_FORMATS:
        try:
            return datetime.strptime(s, fmt)
        except ValueError:
            continue
    try:
        return datetime.fromisoformat(s)
    except ValueError:
        return None


# --- Competition label normalisation -------------------------------------

_COMPETITION_ALIASES = {
    "brasileirao": "Brasileirão Serie A",
    "brasileirão": "Brasileirão Serie A",
    "serie a": "Brasileirão Serie A",
    "série a": "Brasileirão Serie A",
    "serie b": "Brasileirão Serie B",
    "série b": "Brasileirão Serie B",
    "serie c": "Brasileirão Serie C",
    "série c": "Brasileirão Serie C",
    "campeonato brasileiro": "Brasileirão Serie A",
    "copa do brasil": "Copa do Brasil",
    "brazilian cup": "Copa do Brasil",
    "copa do brazil": "Copa do Brasil",
    "libertadores": "Copa Libertadores",
    "copa libertadores": "Copa Libertadores",
    "copa conmebol libertadores": "Copa Libertadores",
}


def canonical_competition(value) -> str | None:
    """Normalise a user-supplied competition name to a canonical label."""
    if value is None:
        return None
    key = strip_accents(str(value)).lower().strip()
    if not key:
        return None
    return _COMPETITION_ALIASES.get(key, str(value).strip())


# --- Player position groups ----------------------------------------------

POSITION_GROUPS = {
    "GK": {"GK"},
    "DEF": {"CB", "RCB", "LCB", "LB", "RB", "LWB", "RWB"},
    "MID": {"CDM", "LDM", "RDM", "CM", "LCM", "RCM", "CAM", "LAM", "RAM",
            "LM", "RM"},
    "FWD": {"ST", "LS", "RS", "LW", "RW", "LF", "RF", "CF"},
}


def position_group(position) -> str | None:
    """Map a granular FIFA position code to a broad group (GK/DEF/MID/FWD)."""
    if position is None or str(position).strip() == "":
        return None
    pos = str(position).strip().upper()
    for group, codes in POSITION_GROUPS.items():
        if pos in codes:
            return group
    return None

File: experiments-local/experiment-17/runs/agent=omp_language=typescript_model=openrouter/z-ai/glm-5.2_tooling=none/rep2-failed/src/normalize.ts
/**
 * brazilian-soccer-mcp / src/normalize.ts
 *
 * Name, date and value normalization.
 *
 * Context block:
 * The six datasets spell team and club names in many incompatible ways:
 *   - state suffixes: "Palmeiras-SP", "Botafogo RJ", "Atletico-MG"
 *   - parenthetical countries (Libertadores): "Nacional (URU)"
 *   - full names: "Atletico Mineiro", "Athletico Paranaense"
 *   - accents present in some files, absent in others: "São Paulo" vs "Sao Paulo"
 *   - spelling variants of the same club: "Atletico-PR" vs "Athletico-PR"
 *
 * To match the SAME club across files without MERGING DISTINCT clubs that share
 * a nickname (Atlético-MG vs Atlético-GO vs Athletico-PR; Botafogo-RJ vs
 * Botafogo-PB; América-MG vs América-RN) we split each name into a `core`
 * (accent/diacritic-free, parenthetical- and state-suffix-stripped, with known
 * full names folded to their short form) and a `state` (from a state column, a
 * trailing suffix, or a parenthetical country). The loader then does a two-pass
 * resolution: a core seen with only one state is "unambiguous" and keyed by its
 * bare core (so "Flamengo-RJ" and "Flamengo" both key to "flamengo"); a core
 * seen with multiple states is "ambiguous" and keyed as `core-state` (so
 * "atletico-mg" and "atletico-go" stay distinct). Query matching is done on the
 * `core`, with an optional state filter for precision.
 */

import type { PositionGroup } from "./types.js";

/** Matches a trailing state/country suffix: "-SP", " - RJ", " RJ" (2 letters). */
const TRAILING_STATE = /[\s-]+([A-Za-z]{2})$/;
/** Matches a parenthetical 2-3 letter country code: "(URU)", "(EQU)". */
const PAREN_COUNTRY = /\(([A-Za-z]{2,3})\)/;

/** Full-name → {core, state?} aliases (lowercased, diacritics-stripped). */
const FULL_NAME_ALIASES: Record<string, { core: string; state?: string }> = {
  "atletico mineiro": { core: "atletico", state: "MG" },
  "atletico goianiense": { core: "atletico", state: "GO" },
  "atletico paranaense": { core: "atletico", state: "PR" },
  "athletico paranaense": { core: "atletico", state: "PR" },
  athletico: { core: "atletico" }, // bare token -> same core as "atletico"
  "america mineiro": { core: "america", state: "MG" },
  "america fc natal": { core: "america", state: "RN" },
  "america f c natal": { core: "america", state: "RN" },
  "sport club do recife": { core: "sport", state: "PE" },
  "sport recife": { core: "sport", state: "PE" },
};
/** Normalize a competition label to a canonical form (e.g. "Serie A" → "Brasileirão"). */
export function canonicalCompetition(raw: string): string {
  const t = (raw ?? "").trim();
  if (/^s[ée]rie\s*a$/i.test(t)) return "Brasileirão";
  if (/^s[ée]rie\s*b$/i.test(t)) return "Brasileirão Série B";
  if (/^s[ée]rie\s*c$/i.test(t)) return "Brasileirão Série C";
  if (/^copa\s*do\s*brasil$/i.test(t)) return "Copa do Brasil";
  if (/libertadores/i.test(t)) return "Libertadores";
  return t;
}

/** Strip diacritics by NFD decomposition + combining-mark removal. */
export function stripDiacritics(s: string): string {
  return s.normalize("NFD").replace(/[\u0300-\u036f]/g, "");
}

/** Remove parenthetical content (including nested). */
function stripParens(s: string): string {
  let out = s;
  let prev: string;
  do {
    prev = out;
    out = out.replace(/\([^()]*\)/g, "");
  } while (out !== prev);
  return out;
}

/** Extract a 2-letter state from a trailing suffix, or null. */
function extractTrailingState(s: string): string | null {
  const m = s.match(TRAILING_STATE);
  return m && m[1] ? m[1].toUpperCase() : null;
}

export interface NameParts {
  /** Canonical short core (lowercased, diacritics stripped), e.g. "atletico". */
  core: string;
  /** Resolved state/country code (uppercase) or null. */
  state: string | null;
  /** Human-readable display name (state suffix/parentheticals trimmed, accents kept). */
  display: string;
}

/**
 * Split a team name into core + state + display.
 * `stateCol` is an explicit state column value (Brazilian files), used in
 * preference to a suffix when present.
 */
export function extractParts(name: string, stateCol?: string | null): NameParts {
  const trimmed = (name ?? "").trim();

  // Parenthetical country (Libertadores), captured before stripping.
  const parenMatch = trimmed.match(PAREN_COUNTRY);
  const parenCountry = parenMatch && parenMatch[1] ? parenMatch[1].toUpperCase() : null;

  // Display name: strip parens + trailing state suffix, keep accents/case.
  let display = stripParens(trimmed).trim();
  const displayState = extractTrailingState(display);
  if (displayState) display = display.replace(TRAILING_STATE, "").trim();
  display = display.replace(/\s+/g, " ").trim();

  // Core: lowercase, diacritics stripped, parens + trailing state removed.
  let base = stripDiacritics(stripParens(trimmed).toLowerCase());
  const trailing = extractTrailingState(base);
  if (trailing) base = base.replace(TRAILING_STATE, "").trim();
  base = base.replace(/\s+/g, " ").trim();
  const alias = FULL_NAME_ALIASES[base];
  const core = alias ? alias.core : base;
  const aliasState = alias?.state ?? null;

  // State resolution: explicit column > trailing suffix > paren country > alias.
  const colState =
    stateCol && /^[A-Za-z]{2}$/.test(stateCol.trim()) ? stateCol.trim().toUpperCase() : null;
  const state = colState ?? trailing ?? parenCountry ?? aliasState;

  return { core, state, display };
}

/**
 * Resolve a club key from core + state, given whether the core is ambiguous
 * (seen with multiple states) and the majority state for fallback.
 */
export function resolveClubKey(
  core: string,
  state: string | null,
  ambiguous: boolean,
  majorityState: string | null,
): string {
  if (!ambiguous) return core;
  const s = state ?? majorityState ?? "X";
  return `${core}-${s.toLowerCase()}`;
}

/** Core portion of a club key (inverse of resolveClubKey for matching). */
export function clubKeyCore(clubKey: string): string {
  const i = clubKey.lastIndexOf("-");
  // Only split off a trailing 2-letter state segment.
  if (i > 0 && clubKey.length - i - 1 <= 3) return clubKey.slice(0, i);
  return clubKey;
}

/**
 * Backwards-compatible bare key (core only). Kept for simple equality use and
 * tests; query matching in the store uses core-based comparison.
 */
export function teamKey(name: string): string {
  return extractParts(name).core;
}

/** Human-readable display name (state suffix/parentheticals trimmed, accents kept). */
export function teamDisplay(name: string): string {
  return extractParts(name).display;
}

/** Parse a date string from any supported format into ISO `YYYY-MM-DD`. */
export function parseDate(raw: string | null | undefined): string | null {
  if (raw == null) return null;
  const s = String(raw).trim();
  if (s === "") return null;

  const iso = s.match(/^(\d{4})-(\d{2})-(\d{2})/);
  if (iso && iso[1] && iso[2] && iso[3]) {
    return `${iso[1]}-${iso[2]}-${iso[3]}`;
  }

  const br = s.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4})/);
  if (br && br[1] && br[2] && br[3]) {
    return `${br[3]}-${br[2].padStart(2, "0")}-${br[1].padStart(2, "0")}`;
  }

  return null;
}

/** Extract a 4-digit season/year from a value (number or numeric string). */
export function parseSeason(value: unknown): number | null {
  if (value == null || value === "") return null;
  const n = typeof value === "number" ? value : Number(String(value).trim());
  return Number.isFinite(n) && n >= 1900 && n <= 2100 ? Math.trunc(n) : null;
}

/** Best-effort integer parse; returns null for blanks/non-numbers. */
export function toInt(value: unknown): number | null {
  if (value == null || value === "") return null;
  if (typeof value === "number") return Number.isFinite(value) ? Math.trunc(value) : null;
  const s = String(value).trim();
  if (s === "") return null;
  const n = Number(s);
  return Number.isFinite(n) ? Math.trunc(n) : null;
}

/** Best-effort float parse; returns null for blanks/non-numbers. */
export function toFloat(value: unknown): number | null {
  if (value == null || value === "") return null;
  if (typeof value === "number") return Number.isFinite(value) ? value : null;
  const s = String(value).trim();
  if (s === "") return null;
  const n = Number(s);
  return Number.isFinite(n) ? n : null;
}

/** Lowercase, diacritic-stripped comparison key for free-text fields. */
export function textKey(s: string): string {
  return stripDiacritics(s).toLowerCase().trim();
}

/** Map a FIFA position code to a coarse position group. */
export function positionGroupOf(position: string): PositionGroup | null {
  const p = position.toUpperCase().trim();
  if (p === "GK") return "goalkeeper";
  if (["CB", "LB", "RB", "LCB", "RCB", "LWB", "RWB"].includes(p)) return "defender";
  if (["CDM", "CM", "CAM", "LM", "RM", "LDM", "RDM", "LCM", "RCM", "LAM", "RAM"].includes(p))
    return "midfielder";
  if (["ST", "LS", "RS", "LW", "RW", "LF", "CF", "RF"].includes(p)) return "forward";
  return null;
}
